### ISSUE TITLE
PowerShell CentOS 7 RPM does not find libgssapi_krb5 on CentOS 7.3

### ISSUE THREAD

**Issue by gorankor** [PowerShell/PowerShell#3329]
Steps to reproduce
------------------
1. Create a test PowerShell script on CentOS 7.3, say "/tmp/test.ps1", obviously change computer name, username and password:

```
$cname = "...";
$UserName = "...";
$PlainPassword = "...";
$Password = $PlainPassword | ConvertTo-SecureString -AsPlainText -Force;
$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $UserName , $Password;
$s = New-PSSession -ComputerName $cname -Credential $Credential -Authentication Negotiate;
if ($s) {
    invoke-command -session $s { Get-Date };
    Remove-PSSession $s;
}
```
2. Run
```
powerShell /tmp/test.ps1
```

Expected behavior
-----------------
It should be able to connect to a remote computer and run the cmdlet.

Actual behavior
---------------
It reports an error:

```
New-PSSession : [...] Connecting to remote server
... failed with the following error message : Could not
start request For more information, see the about_Remote_Troubleshooting Help
topic.
At /tmp/test.ps1:6 char:6
+ $s = New-PSSession -ComputerName $cname -Credential $Credential -Auth ...
+      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (System.Manageme....RemoteRunspace:Re
   moteRunspace) [New-PSSession], PSRemotingTransportException
    + FullyQualifiedErrorId : 1,PSSessionOpenFailed
```

Environment data
----------------
```
Name                           Value
----                           -----
PSVersion                      6.0.0-alpha
PSEdition                      Core
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   3.0.0.0
GitCommitId                    v6.0.0-alpha.17
CLRVersion
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

Workaround
----------------
I was curious what is going on so I looked at strace output and noticed this:

```
13419 open("/opt/microsoft/powershell/6.0.0-alpha.17/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
13419 open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 174
13419 fstat(174, {st_mode=S_IFREG|0644, st_size=31133, ...}) = 0
13419 mmap(NULL, 31133, PROT_READ, MAP_PRIVATE, 174, 0) = 0x7f6b562ab000
13419 close(174)                        = 0
13419 open("/lib64/tls/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
13419 open("/lib64/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
13419 open("/usr/lib64/tls/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
13419 open("/usr/lib64/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)

> $ cd /usr/lib64
> $ ls libgssapi_krb5*
> libgssapi_krb5.so.2  libgssapi_krb5.so.2.2
> 
```
So there's no such file there by default. It works if we create it:

`cd /usr/lib64 && ln -s libgssapi_krb5.so.2.2 libgssapi_krb5.so`

Ideally this should work without manual steps needed.

**Comment by TravisEz13** [comment]
We should do the work in #3349 first (support CentOS 7.3) before we try to fix specific issues on that distribution.

**Comment by joeyaiello** [comment]
Focus on remoting is reserved for subsequent betas, moving out the milestone

**Comment by JamesWTruher** [comment]
@gorankor My assumption is that you are attempting to connect from centos to windows, is that correct?

**Comment by JamesWTruher** [comment]
related to: https://github.com/PowerShell/psl-omi-provider/issues/102

**Comment by giordyb** [comment]
I think the issue is with the krb5-libs package, the symlink gets created after installing krb5-devel.

**Comment by gorankor** [comment]
@JamesWTruher That is correct yes.

**Comment by JamesWTruher** [comment]
@gorankor could you try this again with Beta4?

**Comment by gorankor** [comment]
@JamesWTruher I still see the same issue even with Beta4 on CentOS 7.3 minimal install. Enabling debugging for dynamic loader shows this:

      2535:     file=libgssapi_krb5.so [0];  dynamically loaded by /opt/microsoft/powershell/6.0.0-beta.4/libmi.so [0]
      2535:     find library=libgssapi_krb5.so [0]; searching
      2535:       trying file=/opt/microsoft/powershell/6.0.0-beta.4/libgssapi_krb5.so
      2535:       trying file=/lib64/tls/libgssapi_krb5.so
      2535:       trying file=/lib64/libgssapi_krb5.so
      2535:       trying file=/usr/lib64/tls/libgssapi_krb5.so
      2535:       trying file=/usr/lib64/libgssapi_krb5.so

Do you have this file on your system? What does this show if the answer is yes:

`rpm -qf /usr/lib64/libgssapi_krb5.so`

**Comment by giordyb** [comment]
@gorankor I got it to work on centos by installing the following dependencies

epel-release
krb5-workstation
krb5-devel
gssntlmssp (requires epel-release to be installed beforehand)

**Comment by gorankor** [comment]
@giordyb Sure, I noticed your post, but it's unusual that an application would require "-devel" package like this. I have _krb5-workstation_ and _gssntlmssp_ installed too, but not _krb5-devel_.

**Comment by giordyb** [comment]
@gorankor I have been trying to get Negotiate to work with powershell on linux for the past few weeks and after numerous installs I managed to get it to work only after installing krb5-devel.

I haven't really investigated on WHY it works (the only difference that I could tell was that the symlink to libgssapi_krb5.so got created) , but I agree with you, it should NOT be required.
You can try it yourself: 

- install krb-devel
- either reboot or run ldconfig
- try to connect

you will also see that if you remove the krb5-devel package you won't be able to connect anymore using the negotiate protocol.

**Comment by abhi1202** [comment]
I have tested this and it works for me. I wanted to know how is it different if I use Kerberos authentication mode which fails. 
Is the negotiate authentication mode sends password in secure manner and channel is encrypted ?

**Comment by giordyb** [comment]
@abhi1202 

from the OMI provider docs: 

"A more secure method of authentication uses Secure Protected Negotitation protocol (SPNEGO), which enables negotiation of the security protcol from a selection of options, Currently the only option available with omi is the NT Lan Manager protocol, version 2 (NTLMV2) which is an improvment over Basic auth in two ways.

The password is hashed using an irreversable algorithm, so the password is more secure than Basic.

The client and server support encryption of your data over http connections, so SSL certificates are not required."[link](https://github.com/Microsoft/omi/blob/master/Unix/doc/setup-ntlm-omi.md)

**Comment by JamesWTruher** [comment]
@gorankor I don't have the library /usr/lib64/libgssapi_krb5.so - I do have /usr/lib/libgssapi_krb5.so.2

**Comment by gorankor** [comment]
@JamesWTruher That's the same as me, yet it fails on my system because it can't find the library.

Is your script the same as mine, i.e. does it use **-Authentication Negotiate**?
Can we check where/how it finds it on your system? You'll need to install _strace_:

```
yum install strace
````

Here are my outputs:

```
# cat /tmp/test.ps1
$cname = "...";
$UserName = "...";
$PlainPassword = "...";
$Password = $PlainPassword | ConvertTo-SecureString -AsPlainText -Force;
$Credential = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $UserName , $Password;
$s = New-PSSession -ComputerName $cname -Credential $Credential -Authentication Negotiate;
if ($s) {
    invoke-command -session $s { Get-Date };
    Remove-PSSession $s;
}

# strace -o /tmp/strace.out -s 256 -f powershell /tmp/test.ps1
New-PSSession : [...] Connecting to remote server ... failed with the
following error message : Could not start request For more information, see
the about_Remote_Troubleshooting Help topic.
At /tmp/test.ps1:6 char:6
+ $s = New-PSSession -ComputerName $cname -Credential $Credential -Auth ...
+      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
    + CategoryInfo          : OpenError: (System.Manageme....RemoteRunspace:Re
   moteRunspace) [New-PSSession], PSRemotingTransportException
    + FullyQualifiedErrorId : 1,PSSessionOpenFailed

# grep libgssapi_krb5 /tmp/strace.out
2362  open("/opt/microsoft/powershell/6.0.0-beta.4/libgssapi_krb5.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2362  open("/lib64/libgssapi_krb5.so.2", O_RDONLY|O_CLOEXEC) = 88
2364  open("/opt/microsoft/powershell/6.0.0-beta.4/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2364  open("/lib64/tls/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2364  open("/lib64/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2364  open("/usr/lib64/tls/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2364  open("/usr/lib64/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)

# cd /usr/lib64 && ln -s libgssapi_krb5.so.2.2 libgssapi_krb5.so

# strace -o /tmp/strace.out -s 256 -f powershell /tmp/test.ps1

Friday, July 28, 2017 8:47:25 AM


# grep libgssapi_krb5 /tmp/strace.out
2388  open("/opt/microsoft/powershell/6.0.0-beta.4/libgssapi_krb5.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2388  open("/lib64/libgssapi_krb5.so.2", O_RDONLY|O_CLOEXEC) = 88
2390  open("/opt/microsoft/powershell/6.0.0-beta.4/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2390  open("/lib64/tls/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
2390  open("/lib64/libgssapi_krb5.so", O_RDONLY|O_CLOEXEC) = 91
```

**Comment by TravisEz13** [comment]
Has anyone tried this with the RC?

**Comment by TravisEz13** [comment]
Per @dantraMSFT this is fixed, he is going on vacation, but if you need more information hopefully you can follow-up when he gets back next year.

**Comment by DanTravison** [comment]
This is fixed in the PSRP 1.4.1 release both in the client nuget package and omi provider package.

**Comment by dantraMSFT** [comment]
Additional followup:

I'm finding on CentOS 7.5, I need to install the following to enable NTLM:

epel-release
krb5-workstation
gssntlmssp (requires epel-release to be installed beforehand)

Note that NTLM_USER_FILE is not needed nor is krb5-devel.
